### PR TITLE
doc: Document the CSS architecture

### LINKED ISSUE #1949: DOC: CSS architecture
The CSS (re)architecture is about to be finalised, we need to document this both for internal and external devs.

### PR DESCRIPTION
**Pull request by tobi-or-not-tobi**
This documentation contains the CSS (re)architecture for both components, page layout and global theming. While there's still work in progress, this is a good start to share.

Closes #1949

### CODE REVIEW

File: docs/architecture/css-architecture.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# CSS Architecture
+This document provides a highlevel overview of the CSS architecture of Spartacus.
```
Comment (LTiger14) [inline docs/architecture/css-architecture.md]: high level

File: docs/architecture/css-architecture.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+## Style Library
+Spartacus is provided as a set of standard components, distributed in so-called *npm libraries*. Libraries are used to ensure both extensibility and upgradability at the same time. 
+
+To deliver full flexiblity for styling, all CSS rules are provided in a separate library (`@spartacus/styles`). This make the styles completely optional, configurable and extendible. Customers can extend or replace the standard styles as well as implement their own style rules. 
+
+Additionally, an existing UI framework such as Bootstrap can be used in Spartacus without making it a hardcoded depenendency.
```
Comment (LTiger14) [inline docs/architecture/css-architecture.md]: Depenendency => dependency

File: docs/architecture/css-architecture.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+| 1.0.0   | sparta                    | standard theme for 1.0.0 release
+| 1.0.1   | sparta                    | *patch* release with a sparta theme fix                                                                      |
+| 1.1.0   | sparta, coconut-beta-1    | new features for sparta and new beta release for new coconut theme                                           |
+| 2.0.0   | cocunut (default), sparta | New major version could promote a new theme a the default. The sparta style could become deprated over time. |
+
+
+
+## CSS Technology
+Spartacus is developed with a combination of CSS techniques:
+-  [SASS]() is used as the pre-processing language, like most of todays UI frameworks do
```
Comment (LTiger14) [inline docs/architecture/css-architecture.md]: Missing the links 😉

File: docs/architecture/css-architecture.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+-  [CSS variables]() are used for global theming
+-  [CSS post processing]() is intended to be used to polyfil any of required syntax for older browsers.
+
+### View encapsulation
+
+View encapsulation ensures isolation of styles in single DOM. View encapsulation can be used to ensure that the component style rules from one componet do not interfere with other components. View encapsulation is standardised in a web component architecture, profided by the so-called `shadow-dom`. Applications that do not leverage the shadow-dom there are a few alternatives avaiable:
```
Comment (LTiger14) [inline docs/architecture/css-architecture.md]: View encapsulation can be used to ensure that the component style rules from one `component` do not interfere with other components. View encapsulation is `standardized` in a web component architecture, profided by the so-called shadow-dom.

File: docs/architecture/css-architecture.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+-  [SASS]() is used as the pre-processing language, like most of todays UI frameworks do
+-  [CSS variables]() are used for global theming
+-  [CSS post processing]() is intended to be used to polyfil any of required syntax for older browsers.
+
+### View encapsulation
+
+View encapsulation ensures isolation of styles in single DOM. View encapsulation can be used to ensure that the component style rules from one componet do not interfere with other components. View encapsulation is standardised in a web component architecture, profided by the so-called `shadow-dom`. Applications that do not leverage the shadow-dom there are a few alternatives avaiable: 
+
+- **Emulated encapsulation**  
+  Angular provides *emulated* view encapsulation by adding a (random) component ID to the generated CSS, so that the CSS rules are tightly coupled ot the component and will not intefere with other components.
```
Comment (LTiger14) [inline docs/architecture/css-architecture.md]: intefere => interfere

File: docs/architecture/css-architecture.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+- **Emulated encapsulation**  
+  Angular provides *emulated* view encapsulation by adding a (random) component ID to the generated CSS, so that the CSS rules are tightly coupled ot the component and will not intefere with other components. 
+- **BEM**  
+  BEM is an older technique that use a very specific class name convention to make the component styles specific to the given element(s).
+- **iFrames**  
+  We only mention this to be complete, but it's not a viable option due to a number of disadvantages including bad user experience.
+
+
+None of the above techniques work for spartacus:
+- Angulars emulated encapsulation cannot be used since the component styles are provided by the style library.
```
Comment (LTiger14) [inline docs/architecture/css-architecture.md]: Angular's

File: docs/architecture/css-architecture.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+- BEM is considered old-fashion and complex. Moreover, a well defined finegrained component architecture doesn't need BEM to encapsulate the styles.
```
Comment (LTiger14) [inline docs/architecture/css-architecture.md]: finegrained => fine-grained

File: docs/architecture/css-architecture.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+```
+
+The `optional` flag ensures that code will not break during build whenever a specific import is not part of the imported styles. 
+
+Spartacus will generate the component styles using a simple function that iterates over all the configured component selectors. 
+
+## Page layout styles
+Global theming and component styles are most important to render components on the page, however the overall layout that orchestraters components on a page is another important style layer. This layer is detailed in the [page layout documentation](./page-layout.md).
```
Comment (LTiger14) [inline docs/architecture/css-architecture.md]: that orchestrates?

File: docs/architecture/css-architecture.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+cx-link {
+    color: var(--cx-g-color-primary);
+}
+```
+
+The CSS variables can be customised on the root of the document or for specific selectors. 
+
+## Component Styles
+Spartacus exist of a large number of components that can be used by customers to build their storefront experience. Whitle commerce becomes a comodity, styling is by default opiniated. Not only the colors and fonts, but also the real estate of components as well as backgrounds, lines, etc.
```
Comment (LTiger14) [inline docs/architecture/css-architecture.md]: Spartacus consists of [...]      would be nicer

File: docs/architecture/css-architecture.md (hunk 8 above)
Comment (tobi-or-not-tobi) [reply docs/architecture/css-architecture.md]: rephrased.

**Pull request merged**